### ISSUE TITLE
Gaussian splatting training iteration impact on final output mesh quality

### ISSUE THREAD

**Issue by MagicJoeXZ** [Anttwo/SuGaR#132]
Hey, 

I have a question regarding to impact of Gaussian splatting training iteration times on mesh quality.

As the README points out, 7000 iterations on Gaussian splatting training is recommended. However the output result looks not smooth enough to me. 

These are the training command I used. I extract 300 images from a video footage
```
python gaussian_splatting/train.py -s /home/ubuntu/home_office/ --iterations 7000 -m /home/ubuntu/home_office/gaussian_splatting/home_office

python train.py -s /home/ubuntu/home_office/ -c /home/ubuntu/home_office/gaussian_splatting/home_office/ -r sdf --high_poly True --postprocess_mesh True --square_size 5
```

<img width="1032" alt="Screenshot 2024-02-05 at 5 22 59 PM" src="https://github.com/Anttwo/SuGaR/assets/158340792/5d5fd686-1db6-41f8-9f00-ae16c7174435">

Will increasing Gaussian splatting training iteration improve the mesh quality?

**Comment by ThomasWarn** [comment]
My experience with SuGaR is limited, but I believe that is likely the case. Additionally what image resolution are you utilizing?

You may be experiencing issues regarding the relatively low amount of photos & the dim indoor environment causing noisier than normal input images.

Regarding training for further iterations on the gaussian splatting side - This is my data from one of my tests. As you can see, I am still getting significant improvements to the splatting model even past 10,000 iterations.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/Thomas/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/Thomas/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">


1,000 | 5,000 | 10,000 | 30,000 | 100,000 | 300,000 | 500,000
-- | -- | -- | -- | -- | -- | --
19.016PSNR | 27.518PSNR | 28.901PSNR | 31.56PSNR | 32.236PSNR | 32.746PSNR | 32.578PSNR



</body>

</html>

**Comment by Anttwo** [comment]
Hello @MagicJoeXZ,

The number of iterations might affect the results, but I don’t think that’s the main issue here. 
Can I ask, what do your capture/training images look like? Did you take a 360° shot around the chair?

**From your image, it looks like the foreground object (the chair) is well reconstructed.** 
However, I agree that the background has a more chaotic geometry. 
This is a common issue: Most of the time, 3D reconstruction algorithms produce poor results for background elements, for two main reasons:
1. The background is usually made of textureless, monochrome surfaces, which create an unsolvable mathematical ambiguity in the inverse projective problem.
2. Unlike foreground objects, we usually have much fewer different viewpoints for the background in training images.

Nevertheless, you have several options to improve your results or renders:
1. The simple but _dumb_ idea: Change the shading and choose an ambient lighting. The geometry of the background is not accurate, but changing the shading will hide those bad bumps 😄 
2. You can try the `--low_poly True` configuration. Actually, using fewer polygons is a natural way to smooth the surface and remove some unnecessary bumps. Moreover, when using fewer polygons, SuGaR removes polygons first in areas where they are less needed (low curvature, _etc._), so the quality of your geometry will not necessarily decrease. On the website and in the paper, there are several scenes reconstructed with the low_poly configuration (such as the white and red knight or the _playroom_ for example). You can also adjust the number of vertices by yourself. For information, when using `--low_poly True`, the target number of vertices is 200,000. When using `train.py`, you can remove `low_poly True` and use the argument `--n_vertices_in_mesh 500_000` instead to adjust the number of vertices to 500,000, for instance. Please refer to the `README.md` file for more details.
3. If you want to keep a high number of vertices but have a smoother surface, you can increase the laplacian regularization factor used in `./sugar_trainers/refine.py`. You just have to increase the value of `surface_mesh_laplacian_smoothing_factor` at line 166. This will increase the surface regularization during refinement.

Finally, you can also try to run SuGaR without the postprocessing. Actually, most of the time, postprocessing is not needed as it is designed to remove some very specific artifacts in SuGaR meshes, that appear only if some very fine objects are not covered enough by your capture. Moreover, the current postprocessing strategy can sometimes remove a small number of triangles that shouldn’t be removed, so it can be dangerous (maybe you’ll notice some very small holes in the mesh). That’s why the postprocessing is deactivated by default. So don’t worry about that, you can remove this option when running SuGaR.

**Comment by Anttwo** [comment]
You're welcome @MagicJoeXZ, I’m always happy to help!

Okay, since you took a 360° shot around the chair, it's expected to get a nice-looking chair but a messy background. If you try to take more images of the background elements in your room, you might get a better background (similar to the playroom scene from the DeepBlending dataset, that SuGaR can reconstruct well).

It depends on what you want to achieve: a 360° capture around a single object is ideal to get a high-quality foreground object, and a comprehensive trajectory that covers the entire scene will give a more balanced geometry for both foreground and background.

Sure, I'm looking forward to your answer!

**Comment by Anttwo** [comment]
Woops, sorry, I made a small mistake in my previous message.
To change the mesh regularization factor, you should not change `surface_mesh_laplacian_smoothing_factor` at line 166 as this is not used haha; you should change `'normal_consistency_factor': 0.1,` at line 160 in `train.py`.

Sorry for that.

**Comment by MagicJoeXZ** [comment]
Hey @Anttwo,

Thanks for all the response!
Sorry I didn't explicitly express my use case. Instead of capture the chair it self, I was trying to capture the entire indoor scene. That's why I pass `-r sdf`

Besides the smoothing tips, do you have any suggestions on how to capture and reconstruction indoor scene?


More screenshots 

<img width="1163" alt="Screenshot 2024-02-14 at 1 44 54 PM" src="https://github.com/Anttwo/SuGaR/assets/158340792/f6156618-8001-44fc-9a8a-50ae87b33782">

<img width="1163" alt="Screenshot 2024-02-14 at 1 45 37 PM" src="https://github.com/Anttwo/SuGaR/assets/158340792/e51bbd76-69ca-4a7b-b82b-eef9aa632faf">

**Comment by Anttwo** [comment]
Hey @MagicJoeXZ,

I think I may have what you need, for both capture and smoothing.
I pushed a small change to `sugar_extractors/coarse_mesh.py` and updated the Tips section of the `README.md` file.

Please refer to this `README.md` file for more details. For example, I added some tips to not only remove holes in the mesh, but also reduce these messy ellipsoidal bumps you have on your surface.

Here is the interesting part from the `README.md` file:


> ### 4. I have holes in my mesh, what can I do?
> 
> If you have holes in your mesh, this means the cleaning step of the Poisson mesh is too aggressive for your scene. You can reduce the treshold `vertices_density_quantile` used for cleaning by modifying line 43 of `sugar_extractors/coarse_mesh.py`. For example, you can change this line from
> ```python
>   vertices_density_quantile = 0.1
> ```
> to
> ```python
>   vertices_density_quantile = 0.
> ```
> 
> ### 5. I have messy ellipsoidal bumps on the surface of my mesh, what can I do?
> 
> Depending on your scene, the default hyperparameters used for Poisson reconstruction may be too fine compared to the size of the Gaussians. Gaussian could then become visible on the mesh, which results in messy ellipsoidal bumps on the surface of the mesh.
> This could happen if the camera trajectory is very close to a simple foreground object, for example.<br>
> To fix this, you can reduce the depth of Poisson reconstruction `poisson_depth` by modifying line 42 of `sugar_extractors/coarse_mesh.py`. <br>
> For example, you can change line 42 from
> ```python
>   poisson_depth = 10
> ```
> to
> ```python
>   poisson_depth = 7
> ```
> You may also try `poisson_depth = 6`, or `poisson_depth = 8` if the result is not satisfying.

**Comment by Anttwo** [comment]
It could also be nice to change the depth of the Poisson reconstruction depending on reconstructing the foreground or the background.
For example, here, your chair in the foreground seems to have less bumps than the background in the scene.
I'm going to try to compute automatically the best Poisson depth depending on the scene, and push it to the code.

**Comment by MagicJoeXZ** [comment]
Thanks @Anttwo for quick response and also update on the readme. I'll try your suggestions and compare the output!

**Comment by MagicJoeXZ** [comment]
Thanks @Anttwo!

1. I pulled the latest change
2. change `vertices_density_quantile = 0`, `poisson_depth = 8`. 
3. Don't do postprocessing
4. Pass high poly as False


The results I have:
1. The results is much smoother which is good
2. But also more blobby, I guess that's because poisson_depth change?
3. Also the background area has some 

<img width="1039" alt="Screenshot 2024-02-15 at 6 22 48 PM" src="https://github.com/Anttwo/SuGaR/assets/158340792/18f00e0d-411c-4bf6-9fb5-5242ba3369b9">
<img width="1039" alt="Screenshot 2024-02-15 at 6 23 02 PM" src="https://github.com/Anttwo/SuGaR/assets/158340792/3231202c-8b68-4f7a-a40e-b0d82a35cff1">
<img width="1039" alt="Screenshot 2024-02-15 at 6 23 08 PM" src="https://github.com/Anttwo/SuGaR/assets/158340792/e57f71ab-91ac-4bb1-9470-157446736405">

<img width="1039" alt="Screenshot 2024-02-15 at 6 23 21 PM" src="https://github.com/Anttwo/SuGaR/assets/158340792/0e01dec0-f116-4256-a3ee-ae911284d453">
